alert_words: Performance improvements in looking for alert_words.

benneyman opened the pull request:
This commit leverages the ahocorasick algorithm to build a set of user_ids
that have their alert_words present in the message. It runs in linear time of
the order of length of the input message as opposed to number of alert_words.
This is after building a ahocorasick Automaton which runs in O(number of alert_words in entire realm)
which is usually cached.

<!-- What's this PR for?  (Just a link to an issue is fine.) -->


**Motivation and Implementation**
Currently the logic for finding alert-words in the message content include iterating over the entire alert-words in the realm multiple times i.e.) once to building `possible_words` and another time to check if the each of the `possible_words` is present in the message content. 

The runtime complexity of this is `O(number of words per user * number of users in a realm)` which keeps going up with increase in the users in the realm. 

Using [Aho-Corasick](https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm) algorithm we will be able to do this with a runtime complexity of `O(length of the message content)` after building the Automaton which takes time similar to building `possible_words` . The length of the message content is more or less fixed and smaller in size as opposed to the number of users which can grow as we scale. 

With is commit we can shift the runtime complexity of finding alert-words in message for the entire realm from depending on the number-of alert-words in the realm to length of the message.

I'm using a open-source python package [pyahocorasick](https://pypi.org/project/pyahocorasick/) with BSD License ( BSD-3-Clause and Public-Domain)

**Benchmarks**

I ran benchmarks by generating different number of alert-words with varying lengths of message content .I have created this large number of alert-words for a few users however this is equivalent to a larger number of users with lesser alert-words since the processing of alert-words happens after we take into account all the alert-words in the realm. 

I plotted the difference between the time taken using the regex implementation and the ahocorasick from my branch [here](https://imgur.com/a/WGTOOOb)

- For < 10 words the existing implementation does slightly better than the ahocorasick implementation (by 0.01s), due to the overhead of building the ahocorasick Automaton.
- The time taken to build the ahocorasick Automaton only takes around 0.7s on average for around 50k alert-words in realm. This slight bump in pre-processing helps us achieve linear time complexity later on. 
 - For all alert-words >10 for all varying lengths of content the ahocorasick implemntation does better than the existing implementation especially as the number of alert-words increases the time taken by the ahocorasick imlpementation remains pretty much the same. 
- We'll be able to extend this implementation to do more rules like matching opening/closing parenthesis, looking for substring etc easily without much changes.   
- Even if the time gains is minimal I feel it makes a significant impact as we add more processors to the markdown.

The code used to plot this is 
```
import matplotlib.pyplot as plt

word_counts = [2, 10, 50, 100, 200, 400, 600, 1000, 2000, 4000, 6000, 10000, 20000, 50000]
content_len = [35, 71, 129, 169, 383, 495]
build_aho = [0.020447, 0.002892, 0.003513, 0.003805, 0.004452, 0.006164, 0.013017, 0.016496, 0.024806, 0.042784, 0.071569, 0.284938, 0.234223, 0.74594]
build_regex = [0.006435, 0.003536, 0.002436, 0.002483, 0.002712, 0.005161, 0.003837, 0.00323, 0.0038, 0.004473, 0.005883, 0.00743, 0.015975, 0.026126]

render_regex = {
    2: [0.005892, 0.001907, 0.001913, 0.0026, 0.003792, 0.004224], 
    10: [0.003345, 0.001687, 0.001803, 0.001988, 0.003009, 0.003597], 
    50: [0.005937, 0.00321, 0.004728, 0.004926, 0.005838, 0.007086], 
    100: [0.010994, 0.004963, 0.004746, 0.005415, 0.012205, 0.012117],
    200: [0.024974, 0.005573, 0.007897, 0.012879, 0.018008, 0.021848], 
    400: [0.042395, 0.009812, 0.01635, 0.0182, 0.033834, 0.042116],
    600: [0.236198, 0.224675, 0.231092, 0.279574, 0.274141, 0.260315], 
    1000: [0.36318, 0.376752, 0.39425, 0.388934, 0.41369, 0.431552],
    2000: [0.722153, 0.733265, 0.7507, 0.8738, 0.889466, 0.856505], 
    4000: [1.477045, 1.460673, 1.608463, 1.608304, 1.657198, 1.708451],
    6000: [2.321693, 2.256089, 2.242533, 2.479063, 2.557636, 2.65806], 
    10000: [3.721901, 3.829667, 3.784397, 3.990301, 4.19365, 4.417512],
    20000: [7.463172, 7.506872, 7.769804, 7.816828, 8.463353, 8.690917], 
    50000: [18.545382, 18.795326, 19.23745, 20.234844, 21.262506, 21.989532]
}

render_aho = {
    2: [0.006329, 0.001723, 0.001735, 0.00195, 0.003017, 0.003363], 
    10: [0.0023, 0.001786, 0.001827, 0.001913, 0.0027, 0.00308], 
    50: [0.002272, 0.001851, 0.001926, 0.002137, 0.002944, 0.00319], 
    100: [0.002269, 0.001889, 0.001822, 0.00203, 0.002684, 0.003316], 
    200: [0.002403, 0.001651, 0.001962, 0.002053, 0.002916, 0.003215], 
    400: [0.002324, 0.00181, 0.001917, 0.002022, 0.002819, 0.003151], 
    600: [0.002432, 0.001528, 0.002005, 0.001911, 0.002929, 0.003176], 
    1000: [0.002458, 0.001748, 0.002057, 0.002246, 0.002642, 0.003164], 
    2000: [0.002259, 0.001574, 0.001721, 0.00166, 0.002294, 0.002789], 
    4000: [0.002199, 0.001641, 0.001668, 0.001788, 0.002589, 0.003142], 
    6000: [0.002635, 0.00148, 0.002044, 0.001891, 0.002638, 0.003037], 
    10000: [0.002695, 0.001518, 0.00167, 0.001917, 0.00254, 0.002788], 
    20000: [0.003481, 0.001554, 0.001851, 0.002014, 0.002651, 0.002988], 
    50000: [0.005321, 0.001839, 0.002014, 0.002219, 0.003049, 0.00341]
}

for i, v in enumerate(build_aho):
    word_count = word_counts[i]
    render_aho[word_count] = [x + v for x in render_aho[word_count]]

for i, v in enumerate(build_regex):
    word_count = word_counts[i]
    render_regex[word_count] = [x + v for x in render_regex[word_count]]

fig, ax = plt.subplots(nrows=4, ncols= 4)
index = 0
for row in ax:
    for col in row: 
        if index < len(word_counts):
            col.set_title('Alert word Count = %s' % word_counts[index])
            col.set_xlabel('message content length')
            col.set_ylabel('time taken')
            col.plot(content_len, render_aho[word_counts[index]], marker = 'o')
            col.plot(content_len, render_regex[word_counts[index]], marker = 'o', color='green')
            index += 1
fig.legend(['ahocorasick', 'regex'])

fig.delaxes(ax[3, 2])
fig.delaxes(ax[3, 3])

plt.subplots_adjust(hspace=1.3, wspace=0.4, left=0.07)
plt.show()

```
The code used to profile the ahocorasick implementation is 
```
def test_time_alert_words(self):
    current_times = defaultdict(lambda: defaultdict(float))
    build_times = []
    word_counts = [2, 10, 50, 100, 200, 400, 600, 1000, 2000, 4000, 6000, 10000, 20000, 50000]
    total_tries = 10
    content_len = []
    for word_count in word_counts:
        alert_words_for_users = {
            'hamlet': self.get_random_alert_words(word_count, 10)
        }  # type: Dict[str, List[str]]
        user_profiles = {}  # type: Dict[str, UserProfile]
        for (username, alert_words) in alert_words_for_users.items():
            user_profile = self.example_user(username)
            user_profiles.update({username: user_profile})
            do_set_alert_words(user_profile, alert_words)
        sender_user_profile = self.example_user('hamlet')
        user_ids = {user_profiles['hamlet'].id}
        msg = Message(sender=sender_user_profile, sending_client=get_client("test"))
        build_time_start = time()
        realm_alert_words_automaton = get_alert_word_automaton(sender_user_profile.realm)
        build_time_end = time()
        build_times.append(round(build_time_end - build_time_start, 6))
        def render(msg: Message, content: str) -> str:
            return render_markdown(msg,
                                content,
                                realm_alert_words_automaton=realm_alert_words_automaton,
                                user_ids={user_profile.id})
        content_strings = [
            "This is to test a empty alert words", "i.e. no user has any alert-words set",
            "And to profile how this piece of code stacks up to others ", "as opposed to regex based implementation",
            "This module is written in C. You need a C compiler installed to compile native CPython extensions. To install:"
            "you can use the Automaton class as a trie. Add some string keys and their associated value to this trie.",
            "Here we associate a tuple of (insertion index, original string) as a value to each key string we add to the trie"
        ]
        content = ""
        for content_string in content_strings:
            content = content + content_string
            render_time = 0
            for tries in range(total_tries):
                start_time = time()
                render(msg, content)
                end_time = time()
                render_time += (end_time - start_time)
            current_times[word_count][len(content)] = round(render_time / total_tries, 6)
    ordered_times = OrderedDict(sorted(current_times.items(), key = lambda k : k[0]))
    for (k, v) in ordered_times.items():
        ordered_times[k] = OrderedDict(sorted(v.items(), key = lambda k : k[0]))
        content_len = list(ordered_times[k].keys())
        ordered_times[k] = list(ordered_times[k].values())
    print(ordered_times.items())
    print(build_times)
    print(content_len)

```
And the regex based implementation(existing)
```
def test_time_alert_words(self):
    current_times = defaultdict(list)
    word_counts = [2, 10, 50, 100, 200, 400, 600, 1000, 2000, 4000, 6000, 10000, 20000, 50000]

    for word_count in word_counts:
        alert_words_for_users = {
            'hamlet': self.get_random_alert_words(word_count, 10)
        }  # type: Dict[str, List[str]]
        user_profiles = {}  # type: Dict[str, UserProfile]
        user_ids = set()  # type: Set[int]
        for (username, alert_words) in alert_words_for_users.items():
            user_profile = self.example_user(username)
            user_profiles.update({username: user_profile})
            do_set_alert_words(user_profile, alert_words)
        sender_user_profile = self.example_user('hamlet')
        user_ids = {user_profiles['hamlet'].id}
        msg = Message(sender=sender_user_profile, sending_client=get_client("test"))
        realm_alert_words = alert_words_in_realm(sender_user_profile.realm)

        def render(msg: Message, content: str) -> str:
            return render_markdown(msg,
                                content,
                                realm_alert_words=realm_alert_words,
                                user_ids={user_profile.id})
        content_strings = [
            "This is to test a empty alert words", "i.e. no user has any alert-words set",
            "And to profile how this piece of code stacks up to others ", "as opposed to regex based implementation",
            "This module is written in C. You need a C compiler installed to compile native CPython extensions. To install:"
            "you can use the Automaton class as a trie. Add some string keys and their associated value to this trie.",
            "Here we associate a tuple of (insertion index, original string) as a value to each key string we add to the trie"
        ]
        content = ""
        for content_string in content_strings:
            content = content + content_string
            start_time = time()
            render(msg, content)
            end_time = time()
            current_times[len(content)].append(end_time - start_time)
    print(sorted(current_times.items()))

```





**Testing Plan:** <!-- How have you tested? -->
- Added multiple unit tests in test_bugdown.py covering all scenarios and accented language.

timabbott commented on zerver/lib/message.py:
```diff
@@ -574,8 +575,8 @@ def render_markdown(message: Message,
-    if realm_alert_words is None:
-        realm_alert_words = dict()
+    if realm_alert_words_automaton is None:
+        realm_alert_words_automaton = ahocorasick.Automaton()
```
Same comment on None here as above; is the right behavior to just pass `None` in this case?  If so, we don't need this block.

benneyman replied:
`do_render_markdown` doens't take `None` for `realm_alert_words_automaton ` or for `realm_alert_words ` earlier. So this block will be required unless we make changes to `do_render_markdown`

timabbott commented on zerver/tests/test_bugdown.py:
```diff
@@ -169,10 +172,13 @@ def bugdown_convert(content: str) -> str:
         content=content,
         message_realm=get_realm('zulip'),
         message=message,
+        realm_alert_words_automaton=empty_automaton
```
Same comment on `None` here.

benneyman replied:
True, `bugdown.convert` can take in a `None` and work just fine.

timabbott commented on zerver/lib/alert_words.py:
```diff
@@ -13,6 +15,21 @@ def alert_words_in_realm(realm: Realm) -> Dict[int, List[str]]:
+    alert_word_automaton  = ahocorasick.Automaton()
+    for (user_id, alert_words) in user_id_with_words.items():
+        for alert_word in alert_words:
+            trimmed_alert_word = alert_word.lower()
```
why is this called `trimmed_alert_word`?  It looks like it's just a lower-case version.  Would be better to call it `alert_word_lower` or something similar.

timabbott commented on zerver/tests/test_bugdown.py:
```diff
@@ -892,6 +898,210 @@ def render(msg: Message, content: str) -> str:
+
+        content = "hello how is this possible how are you doing today"
+        render(msg, content)
+        expected_user_ids = {user_profiles['hamlet'].id, user_profiles['cordelia'].id, user_profiles['iago'].id, user_profiles['prospero'].id, user_profiles['othello'].id}  # type: Set[int]
```
Can you line-wrap this?  Our linters don't enforce line-wrapping on tests, because sometimes it's cleaner to have a long line for a URL or something, but this isn't a case of that :)

timabbott commented on zerver/tests/test_bugdown.py:
```diff
@@ -892,6 +898,210 @@ def render(msg: Message, content: str) -> str:
+        for x in range(num_words):
+            alert_word = ''.join([random.choice(allowed_characters) for i in range(10)])
+            alert_words.append(alert_word)
+        return alert_words
```
We don't allow true randomization in our tests, so that we can't have nondeterministic flakes.  Obviously, these random strings are so long that's unlikely to be an issue, but to fit with those standards, can you add a `random.seed(10)` call here?

timabbott replied:
(Or you could just do e.g. `'x' * 500`?)

benneyman replied:
Yes, makes sense. I added this bit so that the huge list of alert_words for profiling resembles the real world usage of alert-words as close as possible.

timabbott commented on zerver/tests/test_bugdown.py:
```diff
@@ -892,6 +898,210 @@ def render(msg: Message, content: str) -> str:
+        alert_words_for_users = {
+            'hamlet': ['réglementaire', 'une politique', 'une merveille'],
+            'cordelia': ['énormément', 'Prod'],
+            'iago': ['prod'], 'prospero': ['deployment'],
```
Can you add a comment here that iago, cordelia, and othello have strings that don't appear in the text and thus should not match?  It'd make the test a bit more readable.

benneyman commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -1644,22 +1660,17 @@ def run(self, lines: Iterable[str]) -> Iterable[str]:
-            allowed_before_punctuation = "|".join([r'\s', '^', r'[\(\".,\';\[\*`>]'])
-            allowed_after_punctuation = "|".join([r'\s', '$', r'[\)\"\?:.,\';\]!\*`]'])
+            if realm_alert_words_automaton is not None and \
+               realm_alert_words_automaton.kind == ahocorasick.AHOCORASICK:
```
I'm not sure what you have in mind when you say the whitespace looks wrong. I split the line and aligned it until the linter stopped screaming at me :) 

Regd. the check for ahocorasick.AHOCORASICK. the automaton can take three values. https://pyahocorasick.readthedocs.io/en/latest/index.html?highlight=Automaton.kind#module-constants
It's `empty` when we have called `make_automaton` without any keys/value pair, `trie` when there is data but `make_automaton` is not called. `AHOCORASICK` when `make_automaton` has been called and there is data. 
To handle cases where there is no alert_words in the realm we do this check. This is a valid scenario so i feel adding an assert statement wouldn't be necessary.

timabbott replied:
That makes sense, but I think if we just allow passing `None` as  `realm_alert_words_automaton`, this case would stop existing and an assert statement would be appropriate.

benneyman commented on zerver/lib/bugdown/__init__.py:
```diff
@@ -1634,6 +1635,21 @@ def possible_linked_stream_names(content: str) -> Set[str]:
+
+    allowed_before_punctuation = set([' ', '\n', '(', '"', '.', ',', '\'', ';', '[', '*', '`', '>'])
+    allowed_after_punctuation = set([' ', '\n', ')', '",', '?', ':', '.', ',', '\'', ';', ']', '!',
+                                     '*', '`'])
```
It's `'\''` in both the lists. `\` to escape a single `'`

timabbott replied:
Oh, just was skimming; I usually spell that `"'"` in this sort of list :).  SGTM.

benneyman commented on zerver/lib/import_realm.py:
```diff
@@ -256,14 +257,14 @@ def fix_message_rendered_content(realm: Realm,
             # platforms, since they generally don't have an "alert
             # words" type feature, and notifications aren't important anyway.
-            realm_alert_words = dict()  # type: RealmAlertWords
+            realm_alert_words_automaton = ahocorasick.Automaton()  # type: ahocorasick.Automaton
```
`do_render_markdown` doesn't take in a `None` only `render_markdown` does. I believe `do_render_markdown` was used as opposed to `render_markdown` since the `sender` is fetched from a `sender_map`  instead of `get_user_profile_by_id(message.sender_id)` from `render_markdown`. So i decided to go with the flow and created a empty automaton since this looked like a one off case

timabbott replied:
I think we should just make `do_render_markdown` support taking `None` for this argument by just tagging that argument as `Optional[ahocorasick.Automaton]`; if the goal is to not have alert words processing happen, that seems like a cleaner way to communicate that than a half-initialized automaton.

benneyman replied:
This makes sense. automaton having None and having it as something other than ahocorasick would mean the same thing. I have modified `do_render_markdown` to take in a `None`. The automaton would be build only in `alert_words.py` and it would return None from there in case it is not of type Ahocorasick

timabbott commented:
@benneyman it looks like you pushed a new version of this; are your changes ready for review?

GitHub doesn't send a reliable automated notification when you push a new version, so you need to make explicit comments when you're ready for your code to be looked at again.

benneyman commented:
@timabbott Oh I didn't really realize that Github doesn't send notifications. All the review comments you mentioned have been completed. I have rebased it to include changes since the last rebase. Its good to reviewed again

timabbott commented:
Nice, merged, thanks @benneyman!  

(Yeah, GitHub can send notifications for comments and new commits, but never does for pushing an updated version of a commit.  And even with pushing a new commit, it's ambiguous whether you're just pushing partial work to run CI or whatever or are ready for another review).

If you're interested in picking up another project in Zulip, PM me on chat.zulip.org.  This was really nice work and I'd be very happy for you to do further work (either related to markdown processor performance or not).

benneyman commented:
@timabbott  Thanks. I'll ping you there and keep skimming thought the code and issues.
